modules/cluster_estimation/cluster_estimation.py

```diff
@@ -0,0 +1,304 @@
+
+        return True
+    
+    def __sort_by_weights(self, model_output:list[np.ndarray,
+                                                                np.float64,
+                                                                np.float64]) -> list[np.ndarray, np.float64, np.float64]:
+        """
+        Sort input model output list by weights in descending order
+        """
+        return sorted(model_output, key = lambda x:x[1], reverse=True)
+    
+    def __get_distance(self, point, cluster):
+        """
+        Calculates distance between a point and a cluster center
+        """
+        diff_vector = [0,0]
+        diff_vector[0] = point[0] - cluster[0]
+        diff_vector[1] = point[1] - cluster[1]
+        
+        return np.linalg.norm(diff_vector)
+
+    def __convert_detections_to_point(self, detections: "list[DetectionInWorld]") -> list[tuple([float, float])]:
+        """
+        Convert DetectionInWorld input object to a [x,y] position to store
+        """
+        points = []
+        for detection in detections:
+            points.append(tuple([detection.centre[0], detection.centre[1]]))
+
+        return points
+    
+    def __filter_by_points_ownership(self, 
+                                     model_output:list[np.ndarray,
+                                                       np.float64,
+                                                       np.float64]) -> list[np.ndarray,
+                                                                            np.float64,
+                                                                            np.float64]:
+        """
+        Removes any clusters that don't have any points belonging to it.
+
+        PARAMETERS
+        ----------
+
+        RETURNS
+        -------
+        """
+        
+        results = self.__vgmm_model.predict(self.__all_points)
+        valid_clusters = []
+
+        
+        # Filtering by each cluster's point ownership
+        unique_clusters, num_points_per_cluster = np.unique(results, return_counts=True)
+        # Remove empty clusters
+        i = 0
+        for i in range(len(model_output)):
+            if (i in unique_clusters):
+                valid_clusters.append(model_output[i])
+
+
+        return valid_clusters
+
+    def __filter_by_covariances(self,
+                                model_output:list[np.ndarray,
+                                                  np.float64,
+                                                  np.float64]) -> list[np.ndarray,
+                                                                       np.float64,
+                                                                       np.float64]:
+        """
+        Removes any cluster with covariances much higher than lowest covariance value
+        """
+        # python list and not np array, need to loop through manually
+        min_covariance = 1E6
+        for item in model_output:
+            if (min_covariance > item[2]):
```

Comment:
Prefer `item[2] < min_covariance` , as `item[2]` is being compared against the current minimum.